### ISSUE TITLE
ONNX export of torch.autograd.grad

### ISSUE THREAD

**Issue by zhangylch** [pytorch/pytorch#55747]
## 🚀 Feature
The onnx in pytorch can not export a model include a torch.autograd.grad.

## Motivation
In molecular dynamics, machine learning potential (MLP) has been a popular way to accelerate the simulations. In the simulations, not only the output(energy) but also the gradient(atomic forces) are required. Thanks to the great job you have done at Pytorch, I have implemented our MLP with Pytorch and it achieves pretty good performance in accuracy and efficiency in training. However, when I employ torchscript to convert the inference model to a serialized model and deploy it to the c++ production(interface with Lammps for dynamics simulation), the speed in simulation is about 4 to 5 times slower than the original implementation with "fortran". Therefore, I try to export the model with onnx and expect to get an acceleratation. However, I encountered an error as shown in the title. The "torch.autograd.grad" is not supported in the "onnx export".

a simple example:
```
import torch
from torch.nn import Linear,Sequential
class testmodule(torch.nn.Module):
    def __init__(self,nlayer):
        super(testmodule,self).__init__()
        module=[]
        for i in range(nlayer):
            module.append(Linear(20,20))
        self.layers=Sequential(*module)

    def forward(self,x):
        out=self.layers(x)
        return out

class pes(torch.nn.Module):
    def __init__(self):
        super(pes,self).__init__()
        self.module=testmodule(3)

    def forward(self,x):
        out=torch.sum(self.module(x))
        force=torch.autograd.grad([out],[x])[0]
        return out,force

nnmodule=torch.jit.script(pes())
x=torch.ones(10,20)
x.requires_grad=True
out=nnmodule(x)
onnx_module=torch.onnx.export(nnmodule,(x),"test.onnx",input_names=['x'],output_names=['y'],\
example_outputs=out,opset_version=12,dynamic_axes={"x":{0:"batch"},"y":{0:"batch"}})
```

## Pitch

Expect the support of export autograd.grad  and index_add in onnx.export.

## Additional context

![捕获](https://user-images.githubusercontent.com/56628666/114295061-6c436600-9ad5-11eb-899c-5444b5528aad.JPG)

cc @houseroad @spandantiwari @lara-hdr @BowenBao @neginraoof @SplitInfinity

**Comment by neginraoof** [comment]
With an autograd function, we can export those to ONNX by adding a symbolic method to the class. An example you van find here:
https://github.com/rwightman/pytorch-image-models/pull/401/files

I'll check the 'grad' functionality to help on how to add the symbolic here.

**Comment by zylustc** [comment]
> With an autograd function, we can export those to ONNX by adding a symbolic method to the class. An example you van find here:
> https://github.com/rwightman/pytorch-image-models/pull/401/files

Sorry for the late reply and thanks for your kind help, I will try this.

**Comment by neginraoof** [comment]
@zylustc 
Thanks, please note that this example works for subclasses of torch.autograd.Function where you define the forward and backward functionality of the op.
Please check if you can utilize this.

**Comment by garymm** [comment]
In theory I think the [Gradient op](https://github.com/onnx/onnx/blob/main/docs/Operators.md#ai.onnx.preview.training.Gradient) in the op set "ai.onnx.preview.training" may be able to represent the same thing as torch.autograd.grad.

Barriers to doing this:
1. torch.onnx.export currently doesn't have a way to output any ops that are not in the default op set ("ai.onnx").
2. The only ONNX implementation / runtime that I know of that might be able to run this model is ONNX Runtime, but you'd have to use a custom build that includes support for the training op set.

This use case seems rare, so the torch.onnx team will not be prioritizing this, but external contributions are welcome.

**Comment by zhangylch** [comment]
Thanks for your reply. It seems like a long time is needed to have this op available. Recently, I tried to build the inference from scratch using libtorch. Do you think it can achieve a sizeable acceleration compared to torhc.jit?

**Comment by justinchuby** [comment]
We’ve gone ahead and closed this issue because it is **stale**.

If you still believe this issue is relevant, please feel free to reopen the issue and we will triage it as necessary. Please specify in a comment any updated information you may have so that we can address it effectively. We encourage you to try the latest pytorch-preview (nightly) version to see if it has resolved the issue.
